torch/csrc/autograd/autograd.cpp

```diff
@@ -0,0 +1,145 @@
+// TODO: Make the Python API above to just call this C++ API.
+variable_list _make_grads(
+    const variable_list& outputs,
+    const variable_list& grad_outputs) {
+  size_t num_tensors = outputs.size();
+  size_t num_gradients = grad_outputs.size();
+  variable_list new_grads;
+  new_grads.reserve(num_tensors);
+  if (grad_outputs.empty()) {
+    for (const Variable& output : outputs) {
+      if (output.requires_grad()) {
+        TORCH_CHECK(
+            output.numel() == 1,
+            "grad can be implicitly created only for scalar outputs");
+        new_grads.emplace_back(at::ones_like(output));
+      }
+    }
+  } else {
+    TORCH_CHECK(
+        num_tensors == num_gradients,
+        "got %ld tensors and %ld "
+        "gradients",
+        num_tensors,
+        num_gradients);
+    for (size_t i = 0; i < outputs.size(); ++i) {
+      const Variable& output = outputs[i];
+      const Variable& grad_output = grad_outputs[i];
+      if (!grad_output.defined()) {
+        if (output.requires_grad()) {
+          TORCH_CHECK(
+              output.numel() == 1,
+              "grad can be implicitly created only for scalar outputs");
+          new_grads.emplace_back(at::ones_like(output));
+        }
+      } else {
+        // grad output is defined, just append to the new_grads
+        new_grads.emplace_back(grad_output);
+      }
+    }
+  }
+  return new_grads;
+}
+variable_list run_backward(
+    const variable_list& outputs,
+    const variable_list& grad_outputs,
+    bool keep_graph,
+    bool create_graph,
+    const variable_list& inputs,
+    bool allow_unused) {
+  size_t num_tensors = outputs.size();
+  edge_list roots;
+  roots.reserve(num_tensors);
+  for (size_t i = 0; i < num_tensors; i++) {
+    const Variable& output = outputs[i];
+    auto gradient_edge = output.gradient_edge();
+    TORCH_CHECK(
+        gradient_edge.function,
+        "element ", i, " of tensors does not require grad and does not have a grad_fn",
+        i);
+    roots.push_back(std::move(gradient_edge));
+  }
+
+  edge_list output_edges;
+  if (!inputs.empty()) {
+    size_t num_inputs = inputs.size();
+    output_edges.reserve(num_inputs);
+    for (size_t i = 0; i < num_inputs; ++i) {
+      const Variable& input = inputs[i];
+      const auto output_nr = input.output_nr();
+      auto grad_fn = input.grad_fn();
+      if (!grad_fn) {
+        grad_fn = input.try_get_grad_accumulator();
+      }
+      TORCH_CHECK(
+          input.requires_grad(),
+          "One of the differentiated Tensors does not require grad");
+      if (!grad_fn) {
+        output_edges.emplace_back();
+      } else {
+        output_edges.emplace_back(grad_fn, output_nr);
+      }
+    }
+  }
+
+  variable_list grad_inputs = Engine::get_default_engine().execute(
+      roots, grad_outputs, keep_graph, create_graph, output_edges);
+  // check if grad_inputs contains None or not base on the allow_unused flag
+  if (inputs.empty()) {
```

Comment:
Just happened to see this piece of code. It seems not right. `num_inputs` will always be 0.